pkg/cluster/cluster.go

```diff
@@ -289,76 +289,74 @@ func (c *Cluster) Inspect(node string) error {
+			if err != nil {
+				return machines, err
+			}
+			// Set Ports
+			ports := make([]config.PortMapping, 0)
+			for k, v := range inspect.NetworkSettings.Ports {
+				if len(v) < 1 {
+					continue
+				}
+				p := config.PortMapping{}
+				hostPort, _ := strconv.Atoi(v[0].HostPort)
+				p.HostPort = uint16(k.Int())
+				p.ContainerPort = uint16(hostPort)
+				p.Address = v[0].HostIP
+				ports = append(ports, p)
+			}
+			m.spec.PortMappings = ports
+			// Volumes
+			var volumes []config.Volume
+			for _, mount := range inspect.Mounts {
+				v := config.Volume{
+					Type:        string(mount.Type),
+					Source:      mount.Source,
+					Destination: mount.Destination,
+					ReadOnly:    mount.RW,
+				}
+				volumes = append(volumes, v)
+			}
+			m.spec.Volumes = volumes
+			m.spec.Cmd = strings.Join(inspect.Config.Cmd, ",")
+			m.ip = inspect.NetworkSettings.IPAddress
+		}
 	}
 	return
 }
 
-func (c *Cluster) gatherMachinesByAPI(all bool) (machines []*Machine, err error) {
-	cli, err := client.NewEnvClient()
+func (c *Cluster) gatherMachineDetails(name string) (container types.ContainerJSON, err error) {
+	res, err := docker.Inspect(name, "{{json .}}")
 	if err != nil {
-		return []*Machine{}, err
-	}
-
-	args := filters.NewArgs()
-	args.Add("label", "works.weave.owner=footloose")
-	if !all {
-		args.Add("label", "works.weave.cluster="+c.spec.Cluster.Name)
+		return container, err
 	}
-	ctx := context.Background()
-	containers, err := cli.ContainerList(ctx, types.ContainerListOptions{
-		Filters: args,
-	})
+	data := []byte(strings.Trim(res[0], "'"))
+	err = json.Unmarshal(data, &container)
 	if err != nil {
-		return []*Machine{}, err
-	}
-
-	for _, container := range containers {
-		m := Machine{}
-		spec := config.Machine{}
-		m.name = container.Names[0]
-		inspect, err := cli.ContainerInspect(ctx, container.ID)
-		if err != nil {
-			return []*Machine{}, err
-		}
-		m.hostname = inspect.Config.Hostname
-		ports := make(map[int]int)
-		for _, p := range container.Ports {
-			ports[int(p.PrivatePort)] = int(p.PublicPort)
-		}
-		m.ports = ports
-		spec.Cmd = container.Command
-		spec.Image = container.Image
-		var volumes []config.Volume
-		for _, mount := range container.Mounts {
-			v := config.Volume{
-				Type:        string(mount.Type),
-				Source:      mount.Source,
-				Destination: mount.Destination,
-				ReadOnly:    mount.RW,
-			}
-			volumes = append(volumes, v)
-		}
-		spec.Volumes = volumes
-		m.spec = &spec
-		machines = append(machines, &m)
+		return container, err
 	}
 	return
 }
 
 func (c *Cluster) gatherMachinesByCluster() (machines []*Machine) {
 	for _, template := range c.spec.Machines {
+		initialMachines := make(map[string]*Machine, 0)
+		names := make([]string, 0)
 		for i := 0; i < template.Count; i++ {
 			machine := c.machine(&template.Spec, i)
-			machines = append(machines, machine)
+			names = append(names, machine.name)
+			initialMachines[machine.name] = machine
+		}
+		sort.Strings(names)
```

Comment:
@dlespiau I wonder. This is from 0-i. That will be in order. Also, this is machine name here which is the same for all machines. I... actually don't think there is a need for doing this. I think the reading order is sufficient. And since slices are order guaranteed they should remain the way they were added and read. 

I'll do some manual tests around this hypothesis.